stubs/WebOb/webob/multidict.pyi

```diff
@@ -0,0 +1,114 @@
+_KT = TypeVar("_KT")
+_VT = TypeVar("_VT")
+
+class MultiDict(MutableMapping[_KT, _VT]):
+    @overload
+    def __init__(self, __m: SupportsItems[_KT, _VT], **kwargs: _VT): ...
+    @overload
+    def __init__(self, __m: Iterable[tuple[_KT, _VT]], **kwargs: _VT): ...
+    @overload
+    def __init__(self, **kwargs: _VT) -> None: ...
+    @classmethod
+    def view_list(cls, lst: list[tuple[_KT, _VT]]) -> MultiDict[_KT, _VT]: ...
+    @classmethod
+    def from_fieldstorage(cls, fs: FieldStorage) -> MultiDict[str, str | FieldStorage]: ...
+    def __getitem__(self, key: _KT) -> _VT: ...
+    def __setitem__(self, key: _KT, value: _VT) -> None: ...
+    def add(self, key: _KT, value: _VT) -> None: ...
+    @overload
+    def get(self, key: _KT) -> _VT: ...
+    @overload
+    def get(self, key: _KT, default: _T) -> _VT | _T: ...
+    def getall(self, key: _KT) -> list[_VT]: ...
+    def getone(self, key: _KT) -> _VT: ...
+    def mixed(self) -> dict[_KT, _VT | list[_VT]]: ...
+    def dict_of_lists(self) -> dict[_KT, list[_VT]]: ...
+    def __delitem__(self, key: _KT) -> None: ...
+    def __contains__(self, key: object) -> bool: ...
+    has_key = __contains__
+    def clear(self) -> None: ...
+    def copy(self) -> Self: ...
+    @overload
+    def setdefault(self, key: _KT, default: None = None) -> _VT | None: ...  # type:ignore[misc]
+    @overload
+    def setdefault(self, key: _KT, default: _VT) -> _VT: ...
+    @overload
+    def pop(self, key: _KT) -> _VT: ...
+    @overload
+    def pop(self, key: _KT, default: _T) -> _VT | _T: ...
+    def popitem(self) -> tuple[_KT, _VT]: ...
+    @overload  # type:ignore[override]
+    def update(self, __m: Collection[tuple[_KT, _VT]], **kwargs: _VT) -> None: ...
+    @overload
+    def update(self, **kwargs: _VT) -> None: ...
+    @overload
+    def extend(self, other: SupportsItems[_KT, _VT], **kwargs: _VT) -> None: ...
+    @overload
+    def extend(self, other: SupportsKeysAndGetItem[_KT, _VT], **kwargs: _VT) -> None: ...
+    @overload
+    def extend(self, other: Iterable[tuple[_KT, _VT]], **kwargs: _VT) -> None: ...
+    @overload
+    def extend(self, other: None = None, **kwargs: _VT) -> None: ...
+    def __len__(self) -> int: ...
+    def keys(self) -> Iterator[_KT]: ...  # type:ignore[override]
+    __iter__ = keys
+    def values(self) -> Iterator[_VT]: ...  # type:ignore[override]
+    def items(self) -> Iterator[tuple[_KT, _VT]]: ...  # type:ignore[override]
+
+class GetDict(MultiDict[str, str]):
+    env: WSGIEnvironment
+    @overload
+    def __init__(self, data: SupportsItems[str, str], env: WSGIEnvironment) -> None: ...
+    @overload
+    def __init__(self, data: Iterable[tuple[str, str]], env: WSGIEnvironment) -> None: ...
+    def on_change(self) -> None: ...
+
+class NestedMultiDict(MultiDict[_KT, _VT]):
+    dicts: tuple[MultiDict[_KT, _VT]]
+    def __init__(self, *dicts: MultiDict[_KT, _VT]) -> None: ...
+    def __setitem__(self, key: _KT, value: _VT) -> NoReturn: ...
```

Comment:
I feel we should *not* use NoReturn for cases where a function always throws an exception to indicate it should not be called. `NoReturn` is meant only for functions that a user would expect to not return, like `sys.exit`, and its primary effect is meant to be to make code after it unreachable. I don't think that's what we want for these functions. See python/typing#1043.

In the meantime, I would suggest annotating these functions with their natural return types (e.g., `None`).

Reply:
That's fair, I would have to agree that unreachable errors are not a good way to let people know that they are not allowed to call a certain function/overload.

On `NoVars` I think we can do a little bit better though, by just removing the method altogether and adding an entry in `stubtest_allowlist.txt`. I would prefer to at least get a static error in functions that return `MultiDict | NoVars`. I am a little bit annoyed that WebOb opted for implementing the whole `MutableMultiDict` interface, just so they can give a slightly better error message at runtime if you try to mutate `NoVars` or retrieve a specific key from it (and to allow it to live inside a `NestedMultiDict`).

Neither `NestedMultiDict` nor `NoVars` are a `MutableMapping` (`NoVars` isn't even a `Mapping`, it just supports most of its methods, save for one) but currently both of them would pass as one. I wish they'd clean this mess up, although I understand why they wouldn't at this point, since it would break people's code everywhere they use `isinstance(x, MultiDict)` currently. It would need to be changed to a common abstract base class `MultiMapping` which both `MultiDict` and `NestedMultiDict` inherit from.